src/common/walFilter/versions/recordProcessGPDB7.c

```diff
@@ -0,0 +1,452 @@
+
+        case RM7_SEQ_ID:
+            if (info == XLOG_SEQ_LOG)
+                return (const RelFileNode *) mainData;
+            THROW_FMT(FormatError, "unknown Sequence: %d", info);
+
+        case RM7_BITMAP_ID:
+            switch (info)
+            {
+                case XLOG_BITMAP_INSERT_WORDS:
+                case XLOG_BITMAP_UPDATEWORD:
+                case XLOG_BITMAP_UPDATEWORDS:
+                case XLOG_BITMAP_INSERT_LOVITEM:
+                case XLOG_BITMAP_INSERT_BITMAP_LASTWORDS:
+                case XLOG_BITMAP_INSERT_META:
+                    return (const RelFileNode *) mainData;
+
+                default:
+                    break;
+            }
+            break;
+
+        case RM7_APPEND_ONLY_ID:
+            switch (info)
+            {
+                case XLOG_APPENDONLY_INSERT:
+                case XLOG_APPENDONLY_TRUNCATE:
+                    return (const RelFileNode *) mainData;
+
+                default:
+                    THROW_FMT(FormatError, "unknown Appendonly: %d", info);
+            }
+    }
+
+    return NULL;
+}
+
+static List *
+getRelFileNodes(XLogRecordGPDB7 *const record, PgPageSize pageSize)
+{
+    #define COPY_HEADER_FIELD(_dst, _size) \
+    do                                     \
+    {                                      \
+        if (remaining < _size)             \
+            return NULL;                   \
+        memcpy(_dst, ptr, _size);          \
+        ptr += _size;                      \
+        remaining -= (uint32_t) _size;     \
+    }                                      \
+    while(0)
+
+    FUNCTION_LOG_BEGIN(logLevelTrace);
+        FUNCTION_LOG_PARAM(XLOG_RECORD_GPDB7, record);
+        FUNCTION_LOG_PARAM(INT, (int) pageSize);
+    FUNCTION_LOG_END();
+
+    size_t remaining = record->xl_tot_len - sizeof(XLogRecordGPDB7);
+    /* Decode the headers */
+    size_t datatotal = 0;
+    int maxBlockId = -1;
+    char *ptr = (char *) record;
+    ptr += sizeof(XLogRecordGPDB7);
+
+    RelFileNode *relFileNode = NULL;
+    List *result = lstNewP(sizeof(RelFileNode));
+    uint32_t mainDataSize = 0;
+    while (remaining > datatotal)
+    {
+        uint8 block_id;
+        COPY_HEADER_FIELD(&block_id, sizeof(block_id));
+
+        if (block_id == XLR_BLOCK_ID_DATA_SHORT)
+        {
+            uint8_t size;
+            COPY_HEADER_FIELD(&size, sizeof(size));
+            mainDataSize = size;
+            break;              /* by convention, the main data fragment is
+                                 * always last */
+        }
+        if (block_id == XLR_BLOCK_ID_DATA_LONG)
+        {
+            COPY_HEADER_FIELD(&mainDataSize, sizeof(mainDataSize));
+            break;              /* by convention, the main data fragment is
+                                 * always last */
+        }
+        if (block_id == XLR_BLOCK_ID_ORIGIN)
+        {
+            ptr += sizeof(RepOriginId);
+            remaining -= (uint32_t) sizeof(RepOriginId);
```

Comment:
remaining is declared as size_t
```suggestion
            remaining -= sizeof(RepOriginId);
```